### PR TITLE
[Arrow] Add `ArrowQueryResult`

### PR DESCRIPTION
**Pull request by Tishj**
This PR adds the `ArrowQueryResult` `QueryResult` variation.

TL;DR:
This `QueryResult` variation is a `MaterializedQueryResult` that contains `ArrowArray`s instead of a `ColumnDataCollection`.

### Motivation

What we currently do to produce Arrow data is first in parallel create a `MaterializedQueryResult` through the `PhysicalMaterializedCollector`, and then the fetched chunks get converted into `ArrowArray`s by a single thread.

### Benefits

Now through the `PhysicalArrowCollector` we also parallelize this last conversion step, skipping the need to collect all data into a ColumnDataCollection first.

### Tests

No extra tests are added, instead I've opted to hijack the existing ArrowRoundTrip code.
What the roundtrip test path used to look like:
- query -> MaterializedQueryResult
- wrap the MaterializedQueryResult into a ArrowFactory
- arrow_scan(factory) -> MaterializedQueryResult (this performs the conversion from DataChunk -> ArrowArray)
- compare result to direct execution (query -> MaterializedQueryResult)

What the roundtrip looks like now:
- query -> ArrowQueryResult
- wrap the ArrowQueryResult into a ArrowFactory
- arrow_scan(factory) -> MaterializedQueryResult (this just feeds the pre-created ArrowArrays)
- compare result to direct execution (query -> MaterializedQueryResult)

Essentially the only thing that really changed is where the conversion takes place (before: factory, now: PhysicalArrowCollector)

### Future work

result collectors need 3 variants:
- single threaded for order preservation when the Source does not support batching
- parallel version for non-order preserving queries
- parallel version that is order-preserving when the Source does support batching

That third variation is not added by this PR yet

### CODE REVIEW

**Comment by Tishj** [comment]
@Mytherin could you restart <https://github.com/duckdb/duckdb/actions/runs/9487449704/job/26144115003?pr=12496>
Somehow `pip` broke when installing a package 🙃

**Comment by Tishj** [comment]
This failure is expected <https://github.com/duckdb/duckdb/actions/runs/9516126601/job/26231763863?pr=12496>
I added a benchmark

**Comment by Tishj** [comment]
Perhaps I should already add a benchmark for the batched version, so the follow up PR for that won't have the same `Regression / Regression Tests` failure like this one does

**Comment by Mytherin** [comment]
Thanks! LGTM
